core/src/com/fourinachamber/fortyfive/map/shop/ShopWidget.kt

```diff
@@ -0,0 +1,191 @@
+import onj.value.OnjNamedObject
+import onj.value.OnjObject
+import kotlin.random.Random
+
+class ShopWidget(
+    texture: String,
+    dataFile: String,
+    private val dataFont: Label.LabelStyle,
+    private val dataDragBehaviour: OnjNamedObject,
+    private val maxPerLine: Int,
+    private val widthPercentagePerItem: Float,
+    val screen: OnjScreen,
+) : CustomFlexBox(screen) {
+
+    private val cards: MutableList<Card> = mutableListOf()
+    private val priceTags: MutableList<CustomLabel> = mutableListOf()
+    private lateinit var boughtIndices: MutableList<Int>
+
+    private lateinit var dragAndDrop: DragAndDrop
+
+    private val allCards: MutableList<Card>
+    private val chances: HashMap<String, Float> = hashMapOf()
+
+    init {
+        backgroundHandle = texture
+        val onj = OnjParser.parseFile(dataFile)
+        cardsFileSchema.assertMatches(onj)
+        onj as OnjObject
+        val cardPrototypes = Card.getFrom(onj.get<OnjArray>("cards"), screen) {}
+        allCards = cardPrototypes.map { it.create() }.toMutableList()
+        curShopWidget = this
+    }
+
+
+    fun addItems(
+        seed: Long,
+        boughtIndices: MutableList<Int>,
+        dragAndDrop: DragAndDrop
+    ) {
+        this.boughtIndices = boughtIndices
+        this.dragAndDrop = dragAndDrop
+        val rnd = Random(seed)
+        val nbrOfItems = (5..8).random(rnd)
+        for (i in 0 until nbrOfItems) {
+            if (chances.size == 0) break
+            val cardId = getCardToAddWithChances(rnd)
+            cards.add(allCards[cardId])
+            chances.remove(allCards[cardId].name)
+            val stringLabel = CustomLabel(screen, "${cards.last().price}$", dataFont, false)
+            stringLabel.setFontScale(0.1F)
+            stringLabel.setAlignment(Align.center)
+            priceTags.add(stringLabel)
+            if (boughtIndices.contains(i)) buyCard(i)
+            add(stringLabel)
+            add(cards.last().actor)
+        }
+        cards.forEach {
+            val behaviour = DragAndDropBehaviourFactory.dragBehaviourOrError(
+                dataDragBehaviour.name,
+                dragAndDrop,
+                it.actor,
+                dataDragBehaviour
+            )
+            dragAndDrop.addSource(behaviour)
+        }
+    }
+
+    private fun getCardToAddWithChances(rnd: Random): Int {
+        val maxWeight = chances.map { (_, b) -> b }.sum()
+        val value = (0.0F..maxWeight).random(rnd)
+        var curSum = 0.0
+        for (e in chances) {
+            curSum += e.value
+            if (curSum > value) {
+                return allCards.indexOf(allCards.find { it.name == e.key })
+            }
+        }
+        return chances.size - 1
+    }
+
+    private fun buyCard(
+        i: Int,
+    ) {
+        makeCardUnmovable(i)
+        priceTags[i].text.clear()
+        priceTags[i].text.append("bought")
+        if (i !in boughtIndices) boughtIndices.add(i)
+        layout()
```

Comment:
call invalidate() instead of layout() here

Reply:
if invalidate() is used, the textures break and therefore i used layout()